File: implementations/java/OneMaxGA.java
// One-Max Genetic Algorithm Implementation in Java
// Author: Genetic Algorithm Performance Comparison Project
// Date: November 19, 2025

import java.util.*;

public class OneMaxGA {
    // GA Parameters
    private static final int POPULATION_SIZE = 100;
    private static final int CHROMOSOME_LENGTH = 100;
    private static final int MAX_GENERATIONS = 500;
    private static final double CROSSOVER_RATE = 0.8;
    private static final double MUTATION_RATE = 0.01;
    private static final int TOURNAMENT_SIZE = 3;
    
    private static final Random random = new Random();
    
    // Individual representation as boolean array
    private static class Individual {
        private boolean[] genes;
        
        public Individual(int length) {
            genes = new boolean[length];
            for (int i = 0; i < length; i++) {
                genes[i] = random.nextBoolean();
            }
        }
        
        public Individual(boolean[] genes) {
            this.genes = genes.clone();
        }
        
        public boolean[] getGenes() {
            return genes.clone();
        }
        
        public int getFitness() {
            int count = 0;
            for (boolean gene : genes) {
                if (gene) count++;
            }
            return count;
        }
        
        public Individual copy() {
            return new Individual(this.genes);
        }
    }
    
    /**
     * Initialize a random population of binary individuals.
     */
    private static List<Individual> initializePopulation(int size, int length) {
        List<Individual> population = new ArrayList<>();
        for (int i = 0; i < size; i++) {
            population.add(new Individual(length));
        }
        return population;
    }
    
    /**
     * Tournament selection with pre-computed fitnesses.
     */
    private static Individual tournamentSelection(List<Individual> population, int[] fitnesses, int tournamentSize) {
        Individual best = null;
        int bestFitness = -1;
        
        for (int i = 0; i < tournamentSize; i++) {
            int candidateIndex = random.nextInt(population.size());
            Individual candidate = population.get(candidateIndex);
            int fitness = fitnesses[candidateIndex];
            if (fitness > bestFitness) {
                bestFitness = fitness;
                best = candidate;
            }
        }
        
        return best.copy();
    }
    
    /**
     * Single-point crossover between two parents.
     */
    private static Individual[] singlePointCrossover(Individual parent1, Individual parent2) {
        if (random.nextDouble() > CROSSOVER_RATE) {
            return new Individual[]{parent1.copy(), parent2.copy()};
        }
        
        int crossoverPoint = random.nextInt(CHROMOSOME_LENGTH - 1) + 1;
        boolean[] genes1 = parent1.getGenes();
        boolean[] genes2 = parent2.getGenes();
        
        boolean[] offspring1Genes = new boolean[CHROMOSOME_LENGTH];
        boolean[] offspring2Genes = new boolean[CHROMOSOME_LENGTH];
        
        System.arraycopy(genes1, 0, offspring1Genes, 0, crossoverPoint);
        System.arraycopy(genes2, crossoverPoint, offspring1Genes, crossoverPoint, 
                        CHROMOSOME_LENGTH - crossoverPoint);
        
        System.arraycopy(genes2, 0, offspring2Genes, 0, crossoverPoint);
        System.arraycopy(genes1, crossoverPoint, offspring2Genes, crossoverPoint, 
                        CHROMOSOME_LENGTH - crossoverPoint);
        
        return new Individual[]{new Individual(offspring1Genes), new Individual(offspring2Genes)};
    }
    
    /**
     * Mutate an individual with specified mutation rate.
     */
    private static void mutate(Individual individual, double mutationRate) {
        boolean[] genes = individual.genes;
        for (int i = 0; i < genes.length; i++) {
            if (random.nextDouble() < mutationRate) {
                genes[i] = !genes[i];
            }
        }
    }
    
    /**
     * Main genetic algorithm function.
     * Returns array with [generations, bestFitness].
     */
    public static int[] runGA() {
        // Initialize population
        List<Individual> population = initializePopulation(POPULATION_SIZE, CHROMOSOME_LENGTH);
        
        for (int generation = 1; generation <= MAX_GENERATIONS; generation++) {
            // Evaluate fitness once per generation
            int[] fitnesses = new int[population.size()];
            for (int i = 0; i < population.size(); i++) {
                fitnesses[i] = population.get(i).getFitness();
            }
            
            // Check for optimal solution
            int maxFitness = 0;
            for (int fitness : fitnesses) {
                if (fitness > maxFitness) maxFitness = fitness;
            }
            
            if (maxFitness == CHROMOSOME_LENGTH) {
                return new int[]{generation, maxFitness};
            }
            
            // Create new population
            List<Individual> newPopulation = new ArrayList<>();
            
            while (newPopulation.size() < POPULATION_SIZE) {
                // Selection
                Individual parent1 = tournamentSelection(population, fitnesses, TOURNAMENT_SIZE);
                Individual parent2 = tournamentSelection(population, fitnesses, TOURNAMENT_SIZE);
                
                // Crossover
                Individual[] offspring = singlePointCrossover(parent1, parent2);
                
                // Mutation
                mutate(offspring[0], MUTATION_RATE);
                mutate(offspring[1], MUTATION_RATE);
                
                // Add to new population
                newPopulation.add(offspring[0]);
                if (newPopulation.size() < POPULATION_SIZE) {
                    newPopulation.add(offspring[1]);
                }
            }
            
            population = newPopulation;
        }
        
        // Final evaluation
        int[] finalFitnesses = new int[population.size()];
        for (int i = 0; i < population.size(); i++) {
            finalFitnesses[i] = population.get(i).getFitness();
        }
        int finalMaxFitness = 0;
        for (int fitness : finalFitnesses) {
            if (fitness > finalMaxFitness) finalMaxFitness = fitness;
        }
        
        return new int[]{MAX_GENERATIONS, finalMaxFitness};
    }
    
    /**
     * Run a single GA instance and return execution time in milliseconds.
     */
    public static double benchmarkSingleRun() {
        long startTime = System.nanoTime();
        int[] result = runGA();
        long endTime = System.nanoTime();
        
        return (endTime - startTime) / 1_000_000.0;
    }
}

File: implementations/csharp/OneMaxGA.cs
// One-Max Genetic Algorithm Implementation in C#
// Author: Genetic Algorithm Performance Comparison Project
// Date: November 19, 2025

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

public class OneMaxGA
{
    // GA Parameters
    private const int POPULATION_SIZE = 100;
    private const int CHROMOSOME_LENGTH = 100;
    private const int MAX_GENERATIONS = 500;
    private const double CROSSOVER_RATE = 0.8;
    private const double MUTATION_RATE = 0.01;
    private const int TOURNAMENT_SIZE = 3;

    private static readonly Random random = new Random();

    // Individual representation as boolean array
    public class Individual
    {
        public bool[] Genes { get; set; }

        public Individual(int length)
        {
            Genes = new bool[length];
            for (int i = 0; i < length; i++)
            {
                Genes[i] = random.NextDouble() < 0.5;
            }
        }

        public Individual(bool[] genes)
        {
            Genes = (bool[])genes.Clone();
        }

        public int Fitness => Genes.Count(gene => gene);

        public Individual Copy() => new Individual(Genes);
    }

    /// <summary>
    /// Initialize a random population of binary individuals.
    /// </summary>
    private static List<Individual> InitializePopulation(int size, int length)
    {
        return Enumerable.Range(0, size)
            .Select(_ => new Individual(length))
            .ToList();
    }

    /// <summary>
    /// Tournament selection with pre-computed fitnesses.
    /// </summary>
    private static Individual TournamentSelection(List<Individual> population, int[] fitnesses, int tournamentSize)
    {
        Individual best = null;
        int bestFitness = -1;

        for (int i = 0; i < tournamentSize; i++)
        {
            int candidateIndex = random.Next(population.Count);
            Individual candidate = population[candidateIndex];
            int fitness = fitnesses[candidateIndex];

            if (fitness > bestFitness)
            {
                bestFitness = fitness;
                best = candidate;
            }
        }

        return best.Copy();
    }

    /// <summary>
    /// Single-point crossover between two parents.
    /// </summary>
    private static (Individual, Individual) SinglePointCrossover(Individual parent1, Individual parent2)
    {
        if (random.NextDouble() > CROSSOVER_RATE)
        {
            return (parent1.Copy(), parent2.Copy());
        }

        int crossoverPoint = random.Next(1, CHROMOSOME_LENGTH);

        var offspring1Genes = parent1.Genes.Take(crossoverPoint)
            .Concat(parent2.Genes.Skip(crossoverPoint))
            .ToArray();

        var offspring2Genes = parent2.Genes.Take(crossoverPoint)
            .Concat(parent1.Genes.Skip(crossoverPoint))
            .ToArray();

        return (new Individual(offspring1Genes), new Individual(offspring2Genes));
    }

    /// <summary>
    /// Mutate an individual with specified mutation rate.
    /// </summary>
    private static void Mutate(Individual individual, double mutationRate)
    {
        for (int i = 0; i < individual.Genes.Length; i++)
        {
            if (random.NextDouble() < mutationRate)
            {
                individual.Genes[i] = !individual.Genes[i];
            }
        }
    }

    /// <summary>
    /// Main genetic algorithm function.
    /// Returns tuple with (generations, bestFitness).
    /// </summary>
    public static (int, int) RunGA()
    {
        var population = InitializePopulation(POPULATION_SIZE, CHROMOSOME_LENGTH);

        for (int generation = 1; generation <= MAX_GENERATIONS; generation++)
        {
            // Evaluate fitness once per generation
            var fitnesses = new int[population.Count];
            for (int i = 0; i < population.Count; i++)
            {
                fitnesses[i] = population[i].Fitness;
            }

            int maxFitness = fitnesses.Max();

            if (maxFitness == CHROMOSOME_LENGTH)
            {
                return (generation, maxFitness);
            }

            var newPopulation = new List<Individual>();

            while (newPopulation.Count < POPULATION_SIZE)
            {
                var parent1 = TournamentSelection(population, fitnesses, TOURNAMENT_SIZE);
                var parent2 = TournamentSelection(population, fitnesses, TOURNAMENT_SIZE);

                var (offspring1, offspring2) = SinglePointCrossover(parent1, parent2);

                Mutate(offspring1, MUTATION_RATE);
                Mutate(offspring2, MUTATION_RATE);

                newPopulation.Add(offspring1);
                if (newPopulation.Count < POPULATION_SIZE)
                {
                    newPopulation.Add(offspring2);
                }
            }

            population = newPopulation;
        }

        var finalFitnesses = new int[population.Count];
        for (int i = 0; i < population.Count; i++)
        {
            finalFitnesses[i] = population[i].Fitness;
        }
        int finalMaxFitness = finalFitnesses.Max();
        return (MAX_GENERATIONS, finalMaxFitness);
    }

    /// <summary>
    /// Run a single GA instance and return execution time in milliseconds.
    /// </summary>
    public static double BenchmarkSingleRun()
    {
        var stopwatch = Stopwatch.StartNew();
        var (generations, bestFitness) = RunGA();
        stopwatch.Stop();

        return stopwatch.Elapsed.TotalMilliseconds;
    }
}